src/credential.cpp

```diff
@@ -115,22 +115,57 @@ operator==(const X509Credential& lhs, const X509Credential& rhs)
 ///
 /// UserInfoVCCredential
 ///
-UserInfoVCCredential::UserInfoVCCredential(bytes userinfo_vc_jwt_in)
+UserInfoVCCredential::UserInfoVCCredential(std::string userinfo_vc_jwt_in)
   : userinfo_vc_jwt(std::move(userinfo_vc_jwt_in))
+  , _vc(std::make_shared<hpke::UserInfoVC>(userinfo_vc_jwt))
 {
-  const auto vc = UserInfoVC(to_ascii(userinfo_vc_jwt));
-
-  const auto& pub = vc.public_key();
-  const auto pub_data = pub.sig.serialize(*pub.key);
-  _signature_scheme = tls_signature_scheme(pub.sig.id);
-  _public_key = SignaturePublicKey{ pub_data };
 }
 
 bool
 // NOLINTNEXTLINE(readability-convert-member-functions-to-static)
 UserInfoVCCredential::valid_for(const SignaturePublicKey& pub) const
 {
-  return pub == _public_key;
+  const auto& vc_pub = _vc->public_key();
+  return pub.data == vc_pub.sig.serialize(*vc_pub.key);
+}
+
+bool
+UserInfoVCCredential::valid_from(const PublicJWK& pub) const
+{
+  const auto& sig = _vc->signature_algorithm();
+  if (pub.signature_scheme != tls_signature_scheme(sig.id)) {
+    return false;
+  }
+
+  const auto sig_pub = sig.deserialize(pub.public_key.data);
+  return _vc->valid_from(*sig_pub);
```

Comment:
❓ 

Should this be return `_vc->valid_for()`?  Looks like it is s SignaturePublicKey instead of a PublicJWK.  Also, that would result in infinite recursion.

Reply:
No, `valid_from()` is correct.

This is confusing because we have multiple types of signature public key around.  The type of `*sig_pub` is `Signature::PublicKey`, so we're not recursing.  We're calling `UserInfoVC::valid_from(const Signature::PublicKey&)` from `UserInfoVCCredential::valid_from(const PublicJWK&)`, basically moving down a layer of abstraction.

That said, I don't love the number of layers of abstraction here.  We should probably clean this up in a follow-on.